chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,375 @@
+        INT20 20,
+        INT19 19,
+        INT18 18,
+        INT17 17,
+        INT16 16,
+        INT15 15,
+        INT14 14,
+        INT13 13,
+        INT12 12,
+        INT11 11,
+        INT10 10,
+        INT9 9,
+        INT8 8,
+        INT7 7,
+        INT6 6,
+        INT5 5,
+        INT4 4,
+        INT3 3,
+        INT2 2,
+        INT1 1,
+        NMI 0   // Non-Maskable Interrupt
+    ]
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    fn enable(&self) {
+        pm::enable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(true);
+        debug!("{}", pm::is_clock_enabled(Clock::PBD(PBDClock::EIC)));
+    }
+
+    fn disable(&self) {
+        pm::disable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(false);
+    }
+
+    fn line_enable(&self, line_num: usize) {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
+        // en == WriteOnly
+        match line_num {
+            0 => regs.en.write(Interrupt::INT.val(mask)),
+            1 => regs.en.write(Interrupt::INT.val(mask)),
+            2 => regs.en.write(Interrupt::INT.val(mask)),
+            3 => regs.en.write(Interrupt::INT.val(mask)),
+            4 => regs.en.write(Interrupt::INT.val(mask)),
+            5 => regs.en.write(Interrupt::INT.val(mask)),
+            6 => regs.en.write(Interrupt::INT.val(mask)),
+            7 => regs.en.write(Interrupt::INT.val(mask)),
+            8 => regs.en.write(Interrupt::INT.val(mask)),
+            _ => debug!("not supported!"),
+        }
+
+        self.line_enable_interrupt(line_num);
+    }
+
+    fn line_disable(&self, line_num: usize) {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
+        match line_num {
+            0 => regs.dis.write(Interrupt::INT.val(mask)),
+            1 => regs.dis.write(Interrupt::INT.val(mask)),
+            2 => regs.dis.write(Interrupt::INT.val(mask)),
+            3 => regs.dis.write(Interrupt::INT.val(mask)),
+            4 => regs.dis.write(Interrupt::INT.val(mask)),
+            5 => regs.dis.write(Interrupt::INT.val(mask)),
+            6 => regs.dis.write(Interrupt::INT.val(mask)),
+            7 => regs.dis.write(Interrupt::INT.val(mask)),
+            8 => regs.dis.write(Interrupt::INT.val(mask)),
+            _ => debug!("not supported!"),
+        }
+
+        self.line_disable_interrupt(line_num);
+    }
+
+    fn line_configure(
+        &self,
+        line_num: usize,
+        int_mode: hil::eic::InterruptMode,
+        filter: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        let mask: u32 = 1 << line_num;
```

Comment:
what happends if line_num > 8 here?

could we do `if line_num > 8 { return; }`?